Fix support points on quad for MappingQGeneric and StraightBoundary.

kronbichler opened the pull request:
When using `MappingQGeneric` on a grid that has been created by extrusion of a 2D mesh, the mapping support points on quads were previously always computed through the manifold. This is inconsistent to the support points on cells that only get computed through the manifold if the cast to `Boundary` returns `NULL`, i.e., it is a manifold not derived from the old boundary classes and in particular, not a straight boundary. In the else case, we should compute the interior points as an interpolation of the surrounding points (i.e., the 4 vertices and all points on the lines).

While the mesh was perfectly fine in 2D, I got inverted cells for mapping degree 5 of the attached test case in 3D. Interestingly, the inversion only happened on the surface of the elements (quads), not in the inside where the `add_weighted_interior_points` method does something reasonable (even though two of the faces have bad points):
![grid_distorted](https://cloud.githubusercontent.com/assets/8276200/22477933/3610e93a-e7e8-11e6-8337-c96e7c3f293c.png)

With the patch, the cells look as expected:
![grid_nice](https://cloud.githubusercontent.com/assets/8276200/22477942/41f5f420-e7e8-11e6-9c65-885a4415965b.png)

This is also the mesh I get in 2D. The test results compute the volume of the object and did run into an exception of a negative Jacobian determinant without this patch.

kronbichler commented:
Note: In the attached mesh, I use two layers of cells which are attached to a CylinderManifolds (to resolve the circular cylinder) whereas I want to approach a square (or rectangle) with the outer two layers of cells, so I really want to have a `StraightManifold` there. Note that I could have obtained proper results by manually defining a blending between the inner cylinder form and outer square, but I find this approach much more convenient.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3810,19 +3810,45 @@ add_quad_support_points(const Triangulation<3,3>::cell_iterator &cell,
+                  a[GeometryInfo<3>::vertices_per_cell +
+                    (polynomial_degree-1)*
+                    GeometryInfo<3>::face_to_cell_lines(face_no,line) + i];
+          add_weighted_interior_points (support_point_weights_on_quad, tmp_points);
```
Hm, you get to this point if the manifold object for the face is not a `Boundary`. But it may still be a legitimate manifold that wants to do something with the points. On the other hand, `add_weighted_interior_points` simply interpolates between the values that already exist, without asking the manifold any further.

Doesn't this yield the wrong result if, for example, the top surface of the domain you have is curved and described by a `Manifold` object that is not a `Boundary` object?

kronbichler replied:
> Hm, you get to this point if the manifold object for the face is not a Boundary

It should be the other way around, you do interpolation in case there is a `Boundary` object. I agree that this is not correct. To my defense, this is the same logic as used for the interior points in a cell. I think the underlying problem is more fundamental: We need a way to distinguish whether a cell (or face or line) has actively set a manifold which provides us with points and the "default" case where we want to compute intermediate points by interpolation from the vertices and the lower level points (such as lines which might have a manifold assigned). In other words, rather than using a cast to `Boundary` as we do right now in three places we would need to check for the default manifold (which is `StraightBoundary` right now). @luca-heltai you wrote the central manifold functionality: Does this sound reasonable to you? In case you do not object, I would change the casts along those lines.

bangerth commented on tests/mappings/mapping_q_mixed_manifolds.cc:
```diff
@@ -0,0 +1,201 @@
+// ---------------------------------------------------------------------
+
+
+// Shows the shape functions implemented and computes the area of cells.
```
outdated comment.

drwells commented:
This sounds similar to my strange issues in #3874.

kronbichler commented:
@drwells What a coincidence that we both had issues with the manifold within less than 20 hours and independently of each other... ;-)
I actually had a look at your issue #3874 before posting the PR but it must be slightly different things because yours is a 2D issue whereas mine is strictly 3D. Note that I just want to fix the *unassigned* quad case to behave the same as the interior of a cell (hex).

kronbichler commented:
I marked this WIP since we need to figure out a more robust way for handling the default manifold. (And besides I also need to run the tests on my machine since I expect changed output in at least a dozen file - if not, we test the manifold/high-order mapping case way too little.)

drwells commented:
@kronbichler The common issue seems to be combining two different manifolds (you are right, of course, since my issue is purely 2D).

luca-heltai commented:
The logic behind the Manifold class was to incorporate in the functions that computes interior points (or better: intermediate points) the information from the geometry. 

Now you are splitting this logic according to whether or not a face has a Boundary class attached to it (which should be the case *only* for boundary faces) and reverting to the old behaviour of computing by hand the intermediate points for some specific cases. 

The `add_weighted_interior_points` should *not* act independent on the Manifold class. In fact, `add_weighted_interior_points` and `get_intermediate_points_on_object` *should* in principle return **the exact same points** if the underlying manifold is a FlatManifold (or a StraightBoundary), so I would not change the logic of `add_quad_support_points`, but that of `get_intermediate_points_on_object` when you pass it a quad in 3d. 

Is there a specific reason why you preferred going this way?

kronbichler commented:
> In fact, `add_weighted_interior_points` and `get_intermediate_points_on_object` should in principle return the exact same points if the underlying manifold is a FlatManifold

They do not, which you could consider either a bug or a feature. The `add_weighted_interior_points` does an interpolation not only from the four vertices (in 2D) but also all the points on a line. Thus, in case we have a manifold (or boundary) attached to a line, the inner points get adjusted. This places the points inside the quad such that they match with the boundary. The algorithm in the first picture above places the points according to the `get_intermediate_points_on_object` function of `FlatManifold` (the default), which conflict with the points set on the line that gets a `CylinderManifold` (curved). The exact same problem would bite you when you set a curved manifold on a surface but not inside elements. If you do not have too many elements and at sufficiently high order, the straight manifold in the interior of the domain will conflict with the curved manifold on the surface.

To summarize, our current code goes through two two different paths with different results, depending on whether we are a `Boundary` or not for the interior of the cells. Clearly, we need to make this more consistent, for which I see two ways:
- Our default manifold when a user does not set anything is not `FlatManifold` but `FlexibleManifold` that uses `add_weighted_interior_points` to define new points. In case none of the objects (lines, quads, hexes) of an element has another manifold assigned, this will be a flat manifold. This is the only way to produce sensible mappings (at least `MappingQGeneric`) in case a user has set a manifold only at the boundary.
- As soon as a user provides a manifold on any object, he must make sure the manifold set is compatible with `FlatManifold` and *blends* with that manifold in the desired way. In particular, we would need to inform a user that setting a curved (say `SphericalManifold` or `CylinderManifold`) cannot be set only on parts of the objects because a straight line that is set immediately inside an element crosses a curved element boundary.

The second option does not sound feasible to me, so only one of the proposed solutions makes sense. Does anybody see another solution?

kronbichler commented:
>  I would not change the logic of `add_quad_support_points`, but that of `get_intermediate_points_on_object` when you pass it a quad in 3d.

I had a more detailed look into the `get_intermediate_points_on_object` and I do not see how that could be made compatible in an appropriate way. We currently only pass the unit cell `line_support_points` into that function but we would need to take all points of the surrounding lines into account for properly interpolating things. To do that, we would need to rewrite the whole manifold data structures to accept this information (rather than only taking the weights of the vertices as they do now).

Note that this is related to a change introduced in #2698 where I fixed the code to actually use the manifold description in the interior of an cell (or face) in case a manifold is given.

bangerth commented:
So let me try and see whether I understand what it is that is that isn't working here. For simplicity, I'm going to assume that the same situation can be reproduced in 2d:

* You have two adjacent cells
* One cell and the edge that separates the two cells is associated with a manifold that causes it to "bulge out" into the other element.
* The other cell is not associated with any manifold and we compute interior points only taking into account vertices and consequently it does not see the bulging out.

Is this approximately correct?

luca-heltai commented:
Martin, I agree with you. On the other hand, we don't need to go through the `get_intermediate_point_on_object`. I introduced that function to maintain the code as backward compatible as possible, but what is done internally in that function is eventually what you do (with some simplifactions, and taking Manifolds into account) in `add_weighted_interior_points`.

Keep in mind that the Manifold data structure was made agnostic about what you pass to it. It does not require the points to be `vertices`. I would be in favour of using always the same logic, otherwise things will become a nightmare.

If you think it is reasonable to use `add_weighted_interior_points` in the "flexible case", then that should be the default, but instead of arbitrarily deciding where to put the new points, restructure the function so that it uses internally `Manifold.get_new_point(weights, points)`. FlatManifold should return the weighted interior points you are computing now, while other manifolds could do more sensible things, taking into account a smarter "blending" between manifolds.

Does it make sense?

kronbichler commented:
@luca-heltai I have to think a bit about the interface of methods we pass to the mapping. At least the basic issue of the different behavior of quads in 3D should be fixed for the 8.5 release, so I put the "high priority" and "8.5 release" tags on this PR. I will not be able to fix this today but try to do it within the next few days.

kronbichler commented:
@bangerth You are right in your description of the problem. Note however that `MappingQGeneric` with the `add_weighted_interior_points` acts as a flexible blending between the "bulging" part and the straight part for quads in 2D and hexes in 3D. However, it does not do the blending for quads in 3D, which I was trying to fix with this PR. But as @luca-heltai suggests we need to think of the interfaces in general.

bangerth commented:
On 02/03/2017 10:49 AM, Martin Kronbichler wrote:
> @bangerth <https://github.com/bangerth> You are right in your description of
> the problem. Note however that |MappingQGeneric| with the
> |add_weighted_interior_points| acts as a flexible blending between the
> "bulging" part and the straight part for quads in 2D and hexes in 3D. However,
> it does not do the blending for quads in 3D, which I was trying to fix with
> this PR.

Do you think you could come up with a smaller testcase that really only starts 
from two cubes and attaches a mapping to one that deforms it? I've tried to 
wrap my head around your testcase, but it's not easy to follow all of the 
steps and which manifold applies at any given time.

kronbichler commented:
@bangerth I will create a smaller test case. It should already manifest itself on a single cube, so I will do that later today.

kronbichler commented:
@bangerth I now create two pictures from a cube, the first one with the changes in the mapping and the second with the current master branch.
![output_good](https://cloud.githubusercontent.com/assets/8276200/22966006/99343fd4-f360-11e6-84e3-8b839f75313f.png)
![output_bad](https://cloud.githubusercontent.com/assets/8276200/22966009/9d1a7e9c-f360-11e6-960b-51df810a89db.png)

Notice that in the code from the master branch, the "surface" grid on the top and bottom faces of the cube gets its points from the flat manifold and this extends beyond the manifold on the left line surrounding the face, leading to an inverted volume. The interesting thing is that this does only affect the faces in 3D, whereas the volume (interior part of cylinder) gets deformed in a flexible way. The same happens in 2D.

kronbichler commented:
Given the discussion we've had in #3874, the approach to blend a flat manifold with a curved one on a surface through the Laplace vector within a single cell prevents high order convergence. After thinking about the problem for a while, I have the following suggestion:

- Add a flag `use_smoothed_points_on_flat_manifold` to `MappingQGeneric` (and `MappingQ`) to enable/disable the smoothing applied by `add_weighted_interior_points`. We have to clearly document what this smoothing means in terms of high order convergence (you cannot expect more than 3.5 or something like that, depending on the shape of the curved part). In case the flag is set and the manifold can be cast to `FlatManifold`, we do the smoothing on the quads in 2D and the quads and hexes in 3D.

- What should the default value of the flag be? Since we used to smooth the high-order mapping before and all code that was used prior to volume `Manifold` approaches relied on this behavior, we will most likely need to set it to `true`, even though I would prefer having it on `false` to produce inverted volumes (at least for some order) and have a user look up how to avoid it. The *good* solution to avoid the problem is to define a transition manifold that smoothly goes from a curved manifold to a flat manifold and extends over more than one element layer as the mesh is refined. This way one will eventually get the correct convergence order (we tested up to degree 8).

bangerth commented:
>  Notice that in the code from the master branch, the "surface" grid
> on the top and bottom faces of the cube gets its points from the
> flat manifold and this extends beyond the manifold on the left line
> surrounding the face, leading to an inverted volume. The interesting
> thing is that this does only affect the faces in 3D, whereas the
> volume (interior part of cylinder) gets deformed in a flexible way.
> The same happens in 2D.

What are manifold ids of the edges? It seems wrong that the curved face 
apparently takes into account curved edges whereas the straight face 
doesn't (or does it? it's not quite clear to me).

Can you remind what order the mapping is?

For the flat surface, it needs to get 5x5 interior points from the 
(flat) manifold. To do this, it must call the manifold in a way that 
asks the manifold, for each of these 25 points, for an interpolated 
point that weighs some surrounding points. Which surrounding points does 
it use? Apparently not just the four vertices, but also some other 
points. Do you recall which ones these are?

I feel that this latter point is where we should start any fix: it seems 
to me like the mapping is not asking the manifold for new points with 
the right set of surrounding point/weights.

kronbichler commented:
> What are manifold ids of the edges?

All of the four edges in front are on the cylindrical manifold, whereas the top and bottom faces are not.

The order of the mapping is 6 here, but it would be essentially the same on quadratic mappings already.

> Which surrounding points does it [the mapping on the flat manifold] use?

That's where the problem is: For the interior of the cell (hex), it does use all the point in the surrounding, i.e., all 8 vertices, the 12 x 5 points on the lines, and the 6 x 5^2 points on the quads. For the faces, it only uses the four vertices, thus creating apparently flat lines on the top and bottom face but not in the interior.

So in a sense you are right, the mapping is not asking the manifold for enough points and we would need to ask the manifold to provide points from all the surrounding and not just the vertices, as @luca-heltai also suggested. The thing is just that I do not know how to do this for a general manifold. For the flat manifold, we use the Laplace vector that contains appropriate weights for how the position of a point inside an object needs to be "shifted" given a position of the surrounding nodes as a "forcing". One could possibly try to get the weights handed over to the manifold by the weights of the Laplace vector and use the respective surrounding points as `add_weighted_interior_points` does. But at this point I'm not sure whether say a spherical manifold would place all the points exactly according to the sphere as we do right now or whether they would end up in weird positions, destroying convergence.

kronbichler commented:
This is now ready for review again. I think we need to introduce a new flag that lets the user decide of what should happen in case there are incompatible manifolds. I added a lengthy discussion to both `MappingQGeneric` and `MappingQ` that explain what happens. In case you agree on this procedure, I will adapt the tests that are now failing due to the change in the mapping on the face which is definitely needed for consistency:

```
	1341 - fe/jacobians.debug (Failed)
	1342 - fe/jacobians.release (Failed)
	1400 - fe/cell_similarity_dgp_nonparametric_06.debug (Failed)
	1401 - fe/cell_similarity_dgp_nonparametric_06.release (Failed)
	1470 - fe/jacobians_face_fe_field.debug (Failed)
	1471 - fe/jacobians_face_fe_field.release (Failed)
	1584 - fe/cell_similarity_05.debug (Failed)
	1585 - fe/cell_similarity_05.release (Failed)
	1608 - fe/cell_similarity_dgp_monomial_06.debug (Failed)
	1609 - fe/cell_similarity_dgp_monomial_06.release (Failed)
	1670 - fe/cell_similarity_dgp_monomial_05.debug (Failed)
	1671 - fe/cell_similarity_dgp_monomial_05.release (Failed)
	1731 - fe/cell_similarity_dgp_nonparametric_05.debug (Failed)
	1732 - fe/cell_similarity_dgp_nonparametric_05.release (Failed)
	1780 - fe/jacobians_face.debug (Failed)
	1781 - fe/jacobians_face.release (Failed)
	1896 - fe/cell_similarity_06.debug (Failed)
	1897 - fe/cell_similarity_06.release (Failed)
	2035 - multigrid/mg_coarse_01.mpirun=3.debug (Failed)
	2036 - multigrid/mg_coarse_01.mpirun=3.release (Failed)
	2051 - multigrid/step-50_01.mpirun=8.debug (Failed)
	2052 - multigrid/step-50_01.mpirun=8.release (Failed)
	2057 - multigrid/mg_coarse_01.mpirun=1.debug (Failed)
	2058 - multigrid/mg_coarse_01.mpirun=1.release (Failed)
	2091 - multigrid/step-50_01.mpirun=5.debug (Failed)
	2092 - multigrid/step-50_01.mpirun=5.release (Failed)
	2110 - multigrid/step-50_01.mpirun=1.debug (Failed)
	2111 - multigrid/step-50_01.mpirun=1.release (Failed)
	4515 - bits/volume_2.debug (Failed)
	4516 - bits/volume_2.release (Failed)
	4583 - bits/step-10-high-order.debug (Failed)
	4584 - bits/step-10-high-order.release (Failed)
	5774 - hp/project_01_curved_boundary.debug (Failed)
	5775 - hp/project_01_curved_boundary.release (Failed)
	7059 - numerics/project_boundary_01.debug (Failed)
	7060 - numerics/project_boundary_01.release (Failed)
	7153 - numerics/boundaries.debug (Failed)
	7154 - numerics/boundaries.release (Failed)
	7265 - numerics/no_flux_10.debug (Failed)
	7266 - numerics/no_flux_10.release (Failed)
	7277 - numerics/project_01_curved_boundary.debug (Failed)
	7278 - numerics/project_01_curved_boundary.release (Failed)

```
This is the difference of 57 tests failing on my machine and the 26 tests that fail on master according to https://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=5241

drwells commented on include/deal.II/fe/mapping_q_generic.h:
```diff
@@ -81,6 +81,47 @@ template <int,int> class MappingQ;
+ * disable the smoothing, placing the additional points strictly according to
+ * the manifold.
+ *
+ * While this smoothing approach works well for relatively convergence orders
```
I think this is a typo: I don't understand the phrase "for relatively convergence orders".

drwells commented:
Aside from the typo this is a nice description of what is going on here.

kronbichler commented:
We could generalize the above approach as suggested by @luca-heltai and let all manifolds do the proper interpolation. It would be a bit of work and I'm not sure if I'm able to complete it. We would then change the name of the current argument `smooth_support_points_on_flat_manifold` into `use_locations_of_all_surrounding_points` that gets activated once we detect that two different manifold ids appear on a cell. (If the same manifold id is given on all entities of a cell, the approach would be inefficient.) In that case, we would send the list of all surrounding points together will a line of the Laplace matrix `lvs` that we use in `add_weighted_interior_points` to the manifold for interpolation. That could work in the general case and be the best solution. However, I'd like to keep that in a different PR if we agree because it should produce equivalent results to the present PR and I just want to update the results through this PR.

kronbichler commented:
I now adjusted all tests - there were actually only six tests where the output should have change whereas the other 10 tests where bugs in my previous implementation. The reason was that I checked for `StraightBoundary` without realizing that `HyperBallBoundary` is derived from `StraightBoundary`, so to check whether we are actually on the default `StraightBoundary` I had to use `typeid` instead. Honestly, the class hierarchy of boundaries is a real mess, it took me half a day to understand all those details, just reinforcing that #2418 should really be high priority for the 9.0 release.

With this change, the code is working as promised, but as said in my last test we should actually try to pull all the code into the manifold because right now we only have it for StraightManifold.

kronbichler commented:
We need to find a solution to this problem before the release, so setting the high priority flag again.

kronbichler commented:
This needs another review now: I now added the functionality to `MappingQGeneric` to get weighted points on all manifolds. To this end, I restructured the way the support points are generated. The mechanism that decides whether to use multiple points or a 

In order to be efficient, I also added a new query method to the manifold that computes multiple points at once. This is also a really nice performance improvement for `ChartManifold` where `pull_back()` might be an expensive operation. I have applications where `pull_back()` does a Newton iteration. Doing it only once

kronbichler commented:
Note that I'm still running the test suite to find out whether there are test failures. It looks like I still need to fix some tests, but the interface should be reviewed...

luca-heltai commented:
I'll let someone else take a look at it, but this should be merged quickly...

kronbichler commented:
@bangerth do you have comments on this code? If not (and nobody else adds something) I will merge this late Monday/early Tuesday. We need to have it in the release because it fixes the severe bug stated in the beginning, but on the other hand it affects many codes with MappingQ and we should give the users using the latest development sources a chance to run through it before we release.

kronbichler commented:
I should mention that the current state of the PR passes the testsuite. There were adaptions to the output of some 3D tests with `Boundary` objects attached due to my change 17 days ago, but no changes were necessary due to the complete overhaul of the mechanism to find the mapping support points of `MappingQGeneric`.

bangerth commented:
Thanks for working on this. I plan on looking at the patch today or tomorrow -- promised!

bangerth commented on include/deal.II/fe/mapping_q_generic.h:
```diff
@@ -81,8 +81,54 @@ template <int,int> class MappingQ;
  * a manifold object to all cells that can then also be used by the mapping
  * classes for higher order mappings.
  *
+ * <h4>Behavior along curved boundaries and with different manifolds</h4>
```
Rather than duplicating this paragraph here, maybe just link to the one in `MappingQ`?

kronbichler replied:
The question should be whether we want to put the full description in `MappingQ` or in `MappingQGeneric`. I would actually prefer to put it in `MappingQGeneric` because that is the more "sane" mapping in my opinion.

bangerth replied:
Either way is fine. You can have the same section heading in the other class, but just refer to the text in the other place.

bangerth commented on include/deal.II/fe/mapping_q_generic.h:
```diff
@@ -524,29 +571,62 @@ class MappingQGeneric : public Mapping<dim,spacedim>
+   * support point weights from the perimeter to the interior of a hex.
+   *
+   * The table itself contains as many columns as there are vertices on a
+   * particular object (2 for a vertex, 4 for a quad, 8 for a hex) and as many
```
vertex -> line

bangerth commented on include/deal.II/fe/mapping_q_generic.h:
```diff
@@ -524,29 +571,62 @@ class MappingQGeneric : public Mapping<dim,spacedim>
-   * outer unit_support_points, i.e.  unit_support_points on the boundary of
-   * the quad
+   * The table itself contains as many columns as there are surrounding points
+   * to a particular object (2 for a vertex, <code>4 + 4*(degree-1)</code> for
```
vertex -> line

bangerth commented on include/deal.II/fe/mapping_q_generic.h:
```diff
@@ -524,29 +571,62 @@ class MappingQGeneric : public Mapping<dim,spacedim>
+   * particular object (2 for a vertex, 4 for a quad, 8 for a hex) and as many
+   * rows as there are strictly interior points.
+   */
+  std::vector<Table<2,double> > support_point_weights_interior;
```
Is it worth adding to the documentation that this table is only used if the smoothing is off?

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -397,6 +400,33 @@ class Manifold : public Subscriptor
 
   /**
+   * Compute a new set of points around the given points
+   * @p surrounding_points. @p weights is a table with as many columns as
```
around the given points -> that interpolate between the given points

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -397,6 +400,33 @@ class Manifold : public Subscriptor
+   * you may get away by implementing only the get_new_point() function.
+   *
+   * The implementation does not allow for @p surrounding_points and
+   * @p new_points to point to the same vector, so make sure pass different
```
-> so make sure TO pass

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -397,6 +400,33 @@ class Manifold : public Subscriptor
+   *
+   * The implementation does not allow for @p surrounding_points and
+   * @p new_points to point to the same vector, so make sure pass different
+   * objects into the function.
```
out of curiosity, why not? Couldn't we just save the original number of surrounding points, and use that as the upper bound for the loop?

kronbichler replied:
The problem is due to the way I implement `Manifold::get_new_points`: Inside that function, we repeatedly call `get_new_points(surrounding_points, weight)` and of course this does not work if `surrounding_points` keeps increasing in size with every point we append to that array. We could of course put a copy function to the generic `add_new_points` in case the pointers are equal. Do you prefer that?

bangerth replied:
Ah, I see. I'd say don't worry about it and leave as is. You have the assertion in there already, so that's fine.

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -723,6 +753,23 @@ class FlatManifold : public Manifold<dim, spacedim>
 
+  /**
+   * Compute a new set of points around the given points
+   * @p surrounding_points. @p weights is a table with as many columns as
```
use the same language as in the base class

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -723,6 +753,23 @@ class FlatManifold : public Manifold<dim, spacedim>
+   * @p weights.
+   *
+   * For this particular implementation, an interpolation of the
+   * @p surrounding_points according to the @p weights is performed.
```
*all* manifolds interpolate. The difference is that the `FlatManifold` simply takes the (weighted) arithmetic mean of the given points in the Cartesian coordinate system in which they are provided.

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -935,6 +982,33 @@ class ChartManifold : public Manifold<dim,spacedim>
 
   /**
+   * Compute a new set of points around the given points
+   * @p surrounding_points. @p weights is a table with as many columns as
```
same language as in base class

bangerth commented on include/deal.II/grid/manifold.h:
```diff
@@ -935,6 +982,33 @@ class ChartManifold : public Manifold<dim,spacedim>
+   * points from the same surrounding points than separate calls to
+   * get_new_point() in case the pull_back() operation is expensive. Often,
+   * this is indeed the case because pull_back() might involve Newton
+   * iterations or something similar in non-trivial manifolds.
```
I think what you want to say here is that doing multiple points at once helps *reuse* some of the information. You still have to do the pull back for each point separately, it's just that you can (may be able to?) do it more efficiently if you have multiple points at once.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -353,8 +353,9 @@ namespace internal
+        QGaussLobatto<1> line_support_points(data.polynomial_degree+1);
         const TensorProductPolynomials<dim>
-        tensor_pols (Polynomials::generate_complete_Lagrange_basis(data.line_support_points.get_points()));
+        tensor_pols (Polynomials::generate_complete_Lagrange_basis(line_support_points.get_points()));
```
Is this function only called once at the beginning of the mapping? If so, this is fine, but if it is called every time we map something, then setting up the `QGaussLobatto` every time may be expensive.

kronbichler replied:
It is called in the constructor of `InternalData` but now I realize it is also called in `do_transform_real_to_unit_cell` for degree 2 and higher. It won't show up in 3D for sure because setting up all the shape functions is definitely much more expensive but in 2D it's not so clear. I will turn it back in to be sure.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -1086,16 +1086,11 @@ namespace
 
-    // in 1d, there are no quads, so return an empty object
-    if (dim == 1)
-      return loqvs;
```
yes, nice job simplifying this.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -1200,19 +1190,131 @@ namespace
+      return Table<2,double>();
+
+    FE_Q<dim> fe(polynomial_degree);
+    std::vector<Point<dim> > unit_support_points = fe.get_unit_support_points();
```
Setting up a FE object is actually quite expensive. If all you need from it is the support points, can't you get this from just one of the usual quadrature objects? That should be far cheaper than setting up all of the FE specific data structures.

kronbichler replied:
While it is easy to get the points from `QGaussLobatto`, what I'm interested in here is the numbering of the degrees of freedom of an FE_Q element. I can also use `FETools::hierarchic_to_lexicographic_numbering` of course and can change if you want so.

bangerth replied:
Yes, I think that would make me happy. The FEs build so much stuff, and expensive stuff for higher orders, that I think it would be better to bypass this if possible.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3621,82 +3623,70 @@ namespace
-                                         const QGaussLobatto<1>        &line_support_points,
-                                         const TriaIterator &iter,
-                                         std::vector<Point<spacedim> > &points)
+  void add_intermediate_points(const Manifold<dim, spacedim> &manifold,
```
I think the name of the original function was a mistake -- maybe call it `add_interior_points()`?

kronbichler replied:
Sure, I will do.

kronbichler replied:
I deleted this function entirely now because there are no two calls that go into the same path of that function, so it is better to do it inline.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3621,82 +3623,70 @@ namespace
-                                         std::vector<Point<spacedim> > &points)
+  void add_intermediate_points(const Manifold<dim, spacedim> &manifold,
+                               const Table<2,double>         &weight_table,
+                               const TriaIterator            &iter,
```
Also while you're there: call the template argument `TriaIteratorType`.

One could probably avoid the first two template arguments of the function by getting the information from the iterator object. But that's something we can do at a later time.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3621,82 +3623,70 @@ namespace
+            break;
           }
+        for (unsigned int i=0; i<new_points.size(); ++i)
+          points.push_back(new_points[i]);
```
Not important, but you can do this more efficiently by saying
`points.insert (new_points.begin(), new_points.end(), points.end());`

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3621,82 +3623,70 @@ namespace
+    for (unsigned int f=0; f<GeometryInfo<3>::faces_per_cell; ++f)
+      if (&cell->face(f)->get_manifold() != &cell->get_manifold())
+        return false;
+    return true;
```
I don't think you need to general template above if you write this function as
```
template <typename TriaIteratorType>
bool check_identical_manifolds_of_quads(const TriaIteratorType &cell)
￼{
  if (TriaIteratorType::dimension >= 2)
    for (unsigned int f=0; f<GeometryInfo<TriaIteratorType::dimension>::lines_per_cell; ++f)
      if (&cell->face(f)->get_manifold() != &cell->get_manifold())
        return false;
  return true;
}
```
That's because `GeometryInfo<TriaIteratorType::dimension>::lines_per_cell` should be defined for all dimensions.

kronbichler replied:
The problem is that I can't call `get_manifold` on all iterators (like the ones for the vertices), therefore I needed to get a special case for 3D. I will add a comment.

bangerth replied:
OK, do as you see fit.

Out of curiosity, does `->get_manifold()` not compile for vertices, or does it throw an exception? In the loop I wrote out above, I'm only looping over the *lines*, not *faces*, so we never enter the body of the loop. As long as it compiles, that's all that matters -- we'll never execute the body, so will never get any exceptions.

kronbichler replied:
Unfortunately, it does not compile because the respective specialization does not provide that call. We might fix it eventually but that is for another day.

bangerth replied:
Yes, no worries.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3746,7 +3736,9 @@ add_line_support_points (const typename Triangulation<dim,spacedim>::cell_iterat
               line->get_manifold() );
 
-          get_intermediate_points_on_object (manifold, line_support_points, line, line_points);
+          line_points.resize(0);
```
line_points.clear()

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3810,19 +3802,53 @@ add_quad_support_points(const Triangulation<3,3>::cell_iterator &cell,
+          all_manifold_ids_are_equal = false;
+      if (smooth_support_points == false || all_manifold_ids_are_equal ||
+          (dynamic_cast<const Boundary<3,3> *>(&face->get_manifold()) &&
+           std::string(typeid(face->get_manifold()).name()).find("StraightBoundary") ==
```
Ugh :-( But nothing we can do about here.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3810,19 +3802,53 @@ add_quad_support_points(const Triangulation<3,3>::cell_iterator &cell,
+      bool all_manifold_ids_are_equal = true;
+      for (unsigned int l=0; l<GeometryInfo<3>::lines_per_face; ++l)
+        if (&face->line(l)->get_manifold() != &face->get_manifold())
+          all_manifold_ids_are_equal = false;
```
Aren't you currently working on a quad face of a 3d cell? If so, why would the manifolds on lines of this hex that are NOT on the current face matter? I think you should be able to just call the helper function defined above with `face`.

bangerth replied:
Oh, I misread -- you are of course only looking at those. Either way, your helper function should probably work here.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3860,62 +3883,83 @@ MappingQGeneric<dim,spacedim>::
+            all_manifold_ids_are_equal = false;
+      if (dim == 3)
+        if (check_identical_manifolds_of_quads(cell) == false)
+          all_manifold_ids_are_equal = false;
```
I think this whole block deserves a helper function that is generic across all dimensions. It's ok if you don't want to do this right now -- just open a PR and I'll eventually replace the code with something I may put into `TriaAccessor`.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3860,62 +3883,83 @@ MappingQGeneric<dim,spacedim>::
-          add_weighted_interior_points (support_point_weights_on_hex, a);
-        break;
-      }
+      switch (dim)
```
this is not wrong but hard to read -- I was expecting an `else` in front of this line. It is unnecessary because you're `return`ing from the previous block, but for readability it may be nicer to enclose this block in an `else` branch anyway.

bangerth commented on source/fe/mapping_q_generic.cc:
```diff
@@ -3860,62 +3883,83 @@ MappingQGeneric<dim,spacedim>::
+          add_quad_support_points (cell, a);
+
+          // then compute the interior points. decide whether to choose
+          // the smoothed variant that include all points and goes directly
```
include -> includes

bangerth commented on source/grid/manifold.cc:
```diff
@@ -122,7 +141,8 @@ get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
 
-      p = get_intermediate_point(p, surrounding_points[permutation[i]],1.0 - weight );
+      if (weight > 0)
+        p = get_intermediate_point(p, surrounding_points[permutation[i]],1.0 - weight );
```
maybe check `weight != 0`?

bangerth commented on source/grid/manifold.cc:
```diff
@@ -131,6 +151,29 @@ get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
+  for (unsigned int row=0; row<weights.size(0); ++row)
+    {
+      std::memcpy(&local_weights[0], &weights(row,0),
+                  surrounding_points.size() * sizeof(double));
```
This is making assumptions about the data layout of `weights`. Would it be significantly less efficient to let the compiler figure this out itself and convert a `for` loop into the equivalent statement?

bangerth commented on source/grid/manifold.cc:
```diff
@@ -518,38 +569,117 @@ get_new_point (const std::vector<Point<spacedim> > &surrounding_points,
+  if (periodicity == Tensor<1,spacedim>())
+    {
+      for (unsigned int i=0; i<surrounding_points.size(); ++i)
+        p += surrounding_points[i] * weights[i];
```
Yes, nice!

bangerth commented:
I think the only question I have is whether the new "smoothing" flag is necessary at all. I would consider the old behavior a bug, not a feature, and would suggest making the changes you made unconditionally, without the backward compatible option. Any reason to *not always* do what you implemented?

luca-heltai commented:
I'm in favour of always doing what Martin implemented.

kronbichler commented on include/deal.II/fe/mapping_q.h:
```diff
@@ -69,6 +69,53 @@ DEAL_II_NAMESPACE_OPEN
+ * inversions with low and medium convergence orders up to approximately three
+ * to four, there is nonetheless an inherent shortcoming because of a
+ * discontinuous mapping that switches from a curved manifold to a flat
+ * manifold within one layer of elements. This will cause the Jacobian
```
I was a bit unclear here: The problem is not that the Jacobian matrix of the transformation is discontinuous but actually already the transformation itself: If I mix a curved description along a line with a flat description inside the element, the prescribed transformation makes a jump. In terms of the finite element theory (I read sections 2.2 and 3.3 in Strang and Fix), this is problematic because the mapping over element boundaries needs to be **globally** continuous for H^1-conforming problems, also over element boundaries. Even though we smooth the jump over a length `h`, the smoothing occurs in one layer of elements and the underlying transformation is thus not uniformly C^0. Thus, the theory cannot guarantee optimal convergence.

bangerth replied:
Yes, this is the bug: "If I mix a curved description along a line with a flat description inside the element, the prescribed transformation makes a jump." But once you fixed this, at least within (the closure, including boundary of) a cell, the mapping is continuous. For it to not be uniformly in C^0, the Jacobian of the mapping needs to blow up in one cell, but remain bounded in its neighbor, right? Is that really what is happening?

kronbichler replied:
No, the Jacobian is not blowing up. But the finite element theory (at least the way I understood Strang and Fix) suggests that the mapping `F` should be independent of `h`. If we do smoothing, we are not independent of `h` anymore. We do switch from a curved surface to a straight surface within a length `h`, so even though the curvature of the curve is only `O(h^2)` the length `h` apparently is enough to destroy the rates. It does so only at a somewhat higher power, producing only a `O(h^3)` error in the affected elements in the L2 norm (or `O(h^2)` in the energy norm), and half a power more in the whole domain. I tried to include this in the new text, please have a look.

bangerth replied:
Hm, bizarre. Never heard of anything like this. I'll have to ask Guermond next time I see him...

kronbichler commented:
> I would consider the old behavior a bug, not a feature, and would suggest making the changes you made unconditionally, without the backward compatible option.

By old behavior, you mean the interpolation from the vertices only? We can make remove the flag and always go through the interpolation from all points that we have already accumulated except for the case where we have a `Boundary` object that is not `StraightBoundary` and we need to use the old-style calls because they cannot do a proper interpolation. (And we remove those with the next release.)

bangerth commented:
Yes, I think that's a reasonable approach.

kronbichler commented:
Alright, I guess that @luca-heltai was referring to the same choice, so I will do that. We will get slightly different results for non-associative point generation in certain manifolds like `SphericalManifold` but it is definitely not worse than before.

kronbichler commented:
I think I now fixed all comments by @bangerth. I have removed the additional flag `smooth_support_points` and always do that in case we have a `Manifold` object and not a `Boundary` object that does not support the smoothing. I am still running the test suite so I might have to make some additional adjustments.

If nobody objects, I would like to keep the last commit that removes the flag (and fixes the comments in `MappingQGeneric`) separate from the previous ones in order to be able to get back to the flag in case we find it necessary. The new algorithm does produce slightly different points inside the quads in some cases, especially for non-associate manifolds like `SphericalManifold`. That said, the new points are often better.

kronbichler commented:
All tests are passing on my machine.

bangerth commented:
OK, thanks for working this out! I think this is progress, and we will get to simplify the code again once we remove the wretched mess we have with the `Boundary` objects...
